File: DataStructure/Trie/q208-implement-trie.py
# 	Q208. 实现Trie
# 实现一个 TrieNode (前缀树)，包含 insert, search, 和 startsWith 这三个操作。
# 示例:
# TrieNode trie = new TrieNode();
# trie.insert("apple");
# trie.search("apple");   // 返回 true
# trie.search("app");     // 返回 false
# trie.startsWith("app"); // 返回 true
# trie.insert("app");
# trie.search("app");     // 返回 true
# 相关题目
# 634. 单词矩阵623. K步编辑559. 查找树服务527. 序列化Trie473. 单词的添加与查找


class TrieNode:
    def __init__(self):
        self.children = {}
        self.is_word = False


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for c in word:
            if c not in node.children:
                node.children[c] = TrieNode()
            node = node.children[c]  # 指针下移
        node.is_word = True

    # ❤若找到，则返回叶子❤
    def find(self, word: str) -> TrieNode:
        cur = self.root
        for c in word:  # 若找到，则指针下移至孩子
            cur = cur.children.get(c)
            if not cur:  # 若不存在，则返回None，不报错
                return None
        return cur

    def search(self, word: str) -> bool:
        leaf = self.find(word)
        return leaf != None and leaf.is_word

    def startsWith(self, prefix: str) -> bool:
        return self.find(prefix) != None

File: DataStructure/Trie/q208_implement_trie.java
package DataStructure.Trie;


import java.util.PrimitiveIterator;

public class q208_implement_trie {
    private class TrieNode {
        TrieNode[] children = new TrieNode[26];
        boolean isWord = false;
    }

    private class Trie {
        private TrieNode root;
        public Trie() {
            root = new TrieNode();
        }

        public void insert(String word) {
            TrieNode cur = root;
            for (int i = 0; i < word.length(); i++) {
                int idx = word.charAt(i) - 'a'; // 优化：数组代替哈希
                if (cur.children[idx] == null)
                    cur.children[idx] = new TrieNode();
                cur = cur.children[idx]; // 指针下移
            }
            cur.isWord = true; // 叶子处
        }

        private TrieNode find(String word) { // 返回word末尾对应的结点
            TrieNode cur = root;
            char[] chars = word.toCharArray();
            for (char c: chars) {
                int idx = c - 'a'; // 优化：数组代替哈希
                if (cur.children[idx] == null)
                    return null;
                cur = cur.children[idx]; // 指针下移
            }
            return cur;
        }

        public boolean search(String word) {
            TrieNode cur = find(word);
            return cur != null && cur.isWord;
        }

        public boolean startsWith(String prefix) {
            return find(prefix) != null;
        }
    }
}
